### ISSUE TITLE
Customized loss function with variable weights

### ISSUE THREAD

**Issue by monirhaji** [microsoft/CNTK#3731]
Hi everyone, 

I have a customized loss function as below:
def customized_loss(coefficient, loss_hard, loss_rand):
loss = (coefficient + 0.5 )* loss_hard + (1.5 - coefficient) * 2 * loss_rand.

and what i like to have is to a variable coefficient : In fact, I wonder if there is any way to change coeff in each minibatch? e.g., define coeff as :

#i : minibatch number
#num_minibatches_to_train: total number of minibatches

coeff  = Cntk.constant(shape = 32, float(i)/float(num_minibatches_to_train))

I don't know how to pass coeff to the function, as when we define the loss function, it's just symbolic and we can't have access to it anymore.

thanks

**Comment by akamotaco** [comment]
Hi,
You may change constant value after each minibatch.

`
import cntk as C
coeff = C.Constant(1.0)
coeff.value # 1.0

# train a minibatch

new_coeff = 0.99
C.assign(coeff, new_coeff).eval() # update coeff value to 0.99
coeff.value # 0.99
`

**Comment by akamotaco** [comment]
Hi,
You may change constant value after each minibatch.

```
import cntk as C
coeff = C.Constant(1.0)
coeff.value # 1.0

 # loss = (coeff+ 0.5 )* loss_hard + (1.5 - coeff) * 2 * loss_rand.
 # train a minibatch

new_coeff = 0.99
C.assign(coeff, new_coeff).eval() # update coeff value to 0.99
coeff.value # 0.99
```

**Comment by monirhaji** [comment]
Actually I have two different functions for computing loss and training minibatch and I define coeff as a global variable. 
now when I change the coeff variable in training minibatch (similar to what you suggested), the coeff value in loss function remains as the initial value and is not changed anymore.
do you have any idea why it happens?

thanks

**Comment by akamotaco** [comment]
I couldn't catch what you said.
The global coeff has changed by assign function.

```
import cntk as C
import numpy as np

coeff = C.Constant(0.9, name='coeff')
dummy = C.Parameter(1)
dummy1 = C.input(1)

loss_hard = 0.1
loss_rand = 0.2
loss = (coeff+ 0.5 )* loss_hard + (1.5 - coeff) * 2 * loss_rand
print(loss.eval()) # 0.38

model = coeff*(1-dummy)
minibatch_loss = C.reduce_mean(C.square(model-dummy1))

learner = C.sgd([dummy],0.0)
trainer = C.Trainer(model,(minibatch_loss,None),learner)

for i in range(10):
    ret = trainer.train_minibatch({dummy1:np.array([1])},
        outputs=[coeff])
    print(ret[1])
    C.assign(coeff,coeff-coeff/10).eval()

print(loss.eval()) # 0.38 to 0.5558568
```

**Comment by monirhaji** [comment]
thank you for the response. But what I meant is like this:
coeff = C.Constant(0.9, name='coeff')

def loss():
  global coeff
   loss = (coeff+ 0.5 )* loss_hard + (1.5 - coeff) * 2 * loss_rand
   return loss

def training_minibatch(trainer):
   learner = C.sgd([dummy],0.0)
   trainer = C.Trainer(model,(loss,None),learner)
   global coeff
   for i in range(10):
       ret = trainer.train_minibatch({dummy1:np.array([1])},
        outputs=[coeff])
        print(ret[1])
       C.assign(coeff,coeff-coeff/10).eval()

**print(loss.eval())**  this still prints 0.38 instead of 0.5558568

does it make sense what I'm looking for?

**Comment by akamotaco** [comment]
I'm sorry, I did not understand your meaning.
If you show a executable sample code, I would like to see it.
